### ISSUE TITLE
[Proposal] The Road to .NET Core Support

### ISSUE THREAD

**Issue by shiftkey** [octokit/octokit.net#1419]
I came back to look at what would be involved with getting Octokit.net running on .NET Core recently now that they had the 1.0 launch, and this is the initial brain dump for what I'm anticipating.
### Pros
- fewer projects to manage
- less project hacks (I'm still not sure how we're getting NSubstitute to work on `netcore50`)
- `.\build FixProjects` is now redundant
- we can now enable `Octokit.Reactive` on all the platforms that `System.Reactive` works on
- can simplify build scripts (`dotnet pack` and `dotnet test` can be used directly)
### Cons
- requires VS2015 Update 3 + additional tooling to contribute (you'll be prompted on first open)
- `project.json` is an interim step, unclear when we'll have to come back to it and port back to MSBuild
### Things To Investigate
- ~~we're using NSubstitute, which depends on [Castle.Core](https://www.nuget.org/packages/Castle.Core/4.0.0-alpha001) which has a pre-release package with .NET Core support - this might impact us running tests on certain platforms~~ we should try out the [pre-release NSubstitute package](https://twitter.com/nsubstitute/status/762118809784229888)
- PCL `Profile259` maps to `netstandard1.0` as per the docs [https://github.com/dotnet/corefx/blob/master/Documentation/architecture/net-platform-standard.md] but `System.Net.Http` has a minimum version of `netstandard1.1` - this might need to be obsoleted during this transition
- Xamarin support is also there (it's `netstandard1.6`) but I'm not familiar enough with how this should transition...
- need to look at what this means for all our Mono build scripts (may simplify things as `dotnet` runs on those platforms too)

**Comment by alfhenrik** [comment]
Looks like `NSubstitute

**Comment by ryangribble** [comment]
Whether in conjunction with this or not, i'd also like to look at moving the builds over to CAKE rather than FAKE...

**Comment by shiftkey** [comment]
@ryangribble here you go #1440

**Comment by mderriey** [comment]
Is anything blocking this issue at the moment? I'd like to help if that's OK.

**Comment by ryangribble** [comment]
No, nothing blocking it... it looks like all the dependencies are supporting dotnet core now

**Comment by mderriey** [comment]
Great. I'll start with something to kick off a discussion around it.

In the meantime, if you think of anything I should be aware of, please let me know. Thanks.

**Comment by hahmed** [comment]
> requires VS2015 Update 3 + additional tooling to contribute (you'll be prompted on first open)

If we can use visual studio code to contribute to ocotkit - that con  would be addressed. 😄

**Comment by radu-matei** [comment]
I would also like to contribute to the .NET Core port, so count me in when it all starts :D

**Comment by shiftkey** [comment]
As a first step, if someone wants to have a look at getting a simple .NET Core project (name it `Octokit.Next` or whatever) building on Appveyor within our existing build scripts. That should open us up to scaling this out so more people can help out.

I've even got a branch [started a while ago which ports one of the projects](https://github.com/octokit/octokit.net/compare/master...shiftkey:what-even-is-dotnet-core) with a `project.json` that was working locally.

**Comment by mderriey** [comment]
sweet, thanks for pointing that out.

what was the rationale behind defining [several](https://github.com/octokit/octokit.net/compare/master...shiftkey:what-even-is-dotnet-core#diff-04cf975081b2ad282e6c41f68f1235e9R25) [netstandard](https://github.com/octokit/octokit.net/compare/master...shiftkey:what-even-is-dotnet-core#diff-04cf975081b2ad282e6c41f68f1235e9R43) [targets](https://github.com/octokit/octokit.net/compare/master...shiftkey:what-even-is-dotnet-core#diff-04cf975081b2ad282e6c41f68f1235e9R61)? I thought if you build for `netstandard1.x`, then any library targeting `netstandard1.y` where `y >= x` can reference it. Totally possible that I'm wrong, though, as I have very limited exposure to .NET Core.

**Comment by shiftkey** [comment]
@mderriey it was mostly me experimenting at the time with the API surface required to target each `netstandard` flavour, but I was also following what we did with [Rx.NET](https://github.com/Reactive-Extensions/Rx.NET/blob/master/Rx.NET/Source/System.Reactive.Core/project.json).

And of course I'll defer to [the documentation](https://github.com/dotnet/corefx/blob/master/Documentation/architecture/net-platform-standard.md#mapping-the-net-platform-standard-to-platforms) around the .NET Platform Standard on this.

**Comment by strich** [comment]
Hey folks. Can I get a summary or where we're at on this? I may be interested in helping out.

**Comment by shiftkey** [comment]
@strich we have an `Octokit.Next` project and tests building in CI. If someone wants to start building against the Octokit source and targeting new platforms they're more than welcome to dive in.

I'd recommend using the `include` options in `buildOptions`, rather than copying the same source into the new folder.

e.g:

```js
...
"buildOptions": {
  "compile": {
    "include": [
      "../Octokit/*.cs"
    ]
  }
}
...
```

**Comment by strich** [comment]
I'm not sure I understand. It would be good if this Next project was at least setup to whatever standard the others are - I have no idea what requirements you have around your build processes, etc. Maybe you could prepare the project, even if it is broken, so people can more readily refactor code?

**Comment by shiftkey** [comment]
@strich Appveyor and Travis are already setup to build and run tests for this Octokit.Next project - you can see this in action [here](https://ci.appveyor.com/project/github-windows/octokit-net/build/0.17.1769#L1259). Aside for the [requirements](https://www.microsoft.com/net/core#windows) for .NET Core you shouldn't need anything else.

[I also started a branch a while ago](https://github.com/octokit/octokit.net/compare/master...shiftkey:what-even-is-dotnet-core) to see what sort of source changes were necessary, but that was before we started the `Octokit.Next` stuff so I'd use that only for reference...

**Comment by shiftkey** [comment]
@strich the project is relatively vanilla because MSBuild does weird things when you mix `csproj` and `xproj`/`project.json` in the same directory. I also wanted to transition things gradually, without interfering with people who wanted to contribute new features and fixes and might not have the latest tools installed.

**Comment by strich** [comment]
Unfortunately I do not know enough about VS projects to figure out how to include the Octokit files in the Next project by reference - IE not copied - and still be able to build and view the project as if it were a normal VS project.

**Comment by mderriey** [comment]
An approach I would use is to get rid of all the `.csproj` files and create `project.json` to replace them. Since `project.json` allows multi-targetting, then the solution would not contain as many different flavours of the Octokit project, but only the main one that would be built for several targets.

You can find several examples of projects that have been ported to .NET Core, but one @shiftkey mentioned earlier, having worked on it, is [Rx.NET](Rx.NET/blob/master/Rx.NET/Source/System.Reactive.Core/project.json).

I hope this helps.

**Comment by radu-matei** [comment]
Updates on `.csproj` and MSBuild.

https://blogs.msdn.microsoft.com/dotnet/2016/10/19/net-core-tooling-in-visual-studio-15/

**Comment by mderriey** [comment]
I'll spend some time working on this very soon.

My intent is to port the main 3 projects to `netstandard1.1`. Then we'll talk about all the other platforms we need to target and how we'll do it. I quite like how [Rx.NET](https://github.com/Reactive-Extensions/Rx.NET/blob/master/Rx.NET/Source/System.Reactive.Core/project.json) was ported, with conditional compilation symbols that map to capabilities more than platforms.

I'll create a PR early to open a discussion

**Comment by ghuntley** [comment]
Yo.

Alright, I had a poke around whilst reviewing @mderriey's progress and there's some stuff bouncing around in my head that needs to be discussed before progressing further.

First up I think the title of this discussion is incorrect; the focus should not be on porting to netcore - it should be a discussion about what which netstandard profile this library should migrate to and how to handle `[Serializable]`

Currently, Octokit has separate projects for `net45`, `netcore45`, `mono`, `monodroid` and `Profile259`. This seems strange as there doesn't appear to be any  solid platform specific reasons for doing this apart from `[Serializable]` jankyness with exceptions. All of these platforms could be 🔥 and converted into a single xproj that generates a netstandard v1.1 library.

From there you can lift the Akavache build/release pipleine; see https://github.com/octokit/octokit.net/issues/1440#issuecomment-261141143

Another thing that needs discussing is the migration to a newer version of `System.Reactive` that supports netstandard and in doing do might warrant bumping the major depending on how you interpret semver as if I remember right there was some strong naming changes + major version of the dependency was changed:

https://docs.microsoft.com/en-us/dotnet/articles/core/versions/index#semantic-versioning

**Comment by mderriey** [comment]
@shiftkey / @ryangribble if/when you have some time, could you please take a look at @ghuntley's comment above? thanks :+1:

**Comment by ghuntley** [comment]
Had a convo with Oren today - he recommends going to `System.Reactive` v3.1.1 as that won't force `net45` consumers into installing the netstandard lib dependency.

**Comment by ryangribble** [comment]
IM happy with the approach suggested by @mderriey in getting the main 3 projects going on netstandard 1.1 first then looking at what to do next.  I'm also :+1: to the Rx approach of the capability based #ifdefs

**Comment by shiftkey** [comment]
@ghuntley the use of `[Serializable]` was more of a correctness thing than anything core to Octokit's behaviour, so don't let that get in the way of anything serious. I think the most interesting dependency is `System.Net.Http` so I think that sets a floor of `netstandard1.1` for which version to target.

I think this would affect existing users of `Profile259` (from memory that's closer to `netstandard1.0`) but I'm okay with deprecating that.

And yes, I'm :thumbsup: on borrowing your work on the pipeline stuff. I hope @ryangribble is already all over that :grin:

I'm also :thumbsup: to bumping to `System.Reactive` as part of this work - that's probably a good point to semver up to 1.x _finally_.

**Comment by DamianEdwards** [comment]
@shiftkey hello sir! Are you currently working on this? Are you moving straight to the new CSPROJ project format as you go (meaning you'll need >=VS2017 RC or latest dotnet CLI builds)?

Also, please note you should drop your reference to Microsoft.Net.Http and instead just depend on NETStandard.Library which brings in System.Net.Http. The Microsoft.Net.Http package is hella old now and won't be moved into the shiny .NET Standard world.

**Comment by ryangribble** [comment]
Hi @DamianEdwards ,  @mderriey is doing most of this work and I've been changing our build script over from FAKE to CAKE. 

As time goes on i think its more and more likely we do jump onto the "new" csproj project system. Last I heard there were still problems with the tests/explorer in vs2017 but things change so quickly that may be old info?

We'll certainly take your advice on the system.net.http stuff 👍

The main outstanding thing for me in terms of actually getting this dnx build out there is that we previously do source linking in the pdb files and it hasn't seemed that any of the available options (sourcelink gitlink etc) are yet supporting dotnet core projects most of them were also not bothering with project.json and waiting till the new csproj format.  This reminds me I need to go and look at where those guys are at now it's been a few more weeks down the track 😀

**Comment by ryangribble** [comment]
Btw see #1503 for where we are at with this work so far

**Comment by ghuntley** [comment]
Heads up AppVeyor doesn't have VS2017 in GA yet. See https://github.com/appveyor/ci/issues/1179

**Comment by mderriey** [comment]
Hi @DamianEdwards , we only reference `Microsoft.Net.Http` for the `net45` TFM. `NETStandard.Library` is used when targeting `netstandard1.1`, as you can see [here](https://github.com/octokit/octokit.net/pull/1503/files#diff-04cf975081b2ad282e6c41f68f1235e9).

I'll leave it to @shiftkey and @ryangribble to decide whether we make the move to the new `csproj` system with VS 2017 RC. As it's available in AppVeyor, it shouldn't too much of an issue.

**Comment by shiftkey** [comment]
@mderriey @ryangribble @DamianEdwards the state of the tooling has been an issue for me, and was one of the reasons I deferring the porting work. Ideally I'd love for this stuff to "just work" for new contributors who stumble upon this, and having to be on bleeding-edge stuff isn't something that everyone is keen to do. 

>  Are you moving straight to the new CSPROJ project format as you go (meaning you'll need >=VS2017 RC or latest dotnet CLI builds)?

Personally, requiring VS2017 when it's still RC doesn't feel great. I know the benefits are there from the tooling side to skip project.json completely, the only concern I have is about how opening these projects in an incompatible version of VS works currently - probably something I can test myself. 

I'm not really actively working on .NET stuff these days, aside from supporting existing projects and helping them with porting wherever I can, so I'm going to defer to the others in this thread - especially @mderriey who has been championing this work - about whether they want to move up to VS2017 sooner.

If the infrastructure is there to support the new csproj changes (and it sounds like Appveyor is close to supporting VS2017 bits), then we can make it clearer to end users to go and grab VS2017 Community if they want to get involved.

> Also, please note you should drop your reference to Microsoft.Net.Http and instead just depend on NETStandard.Library which brings in System.Net.Http. 

Great to hear :sparkles:

**Comment by mderriey** [comment]
> we only reference `Microsoft.Net.Http` for the `net45` TFM.

Did I *really* write this? That's embarassing. Point taken, will remove it altogether.

> Are you moving straight to the new CSPROJ project format as you go (meaning you'll need >=VS2017 RC or latest dotnet CLI builds)?

I tried the one-way upgrade over the WE with the latest VS 2017 RC. Didn't go well as the upgrade didn't work when opening the solution. I'll have another look this week. I know @NickCraver did it for MiniProfiler so I'll have a look at how he got it done.

**Comment by mderriey** [comment]
@shiftkey I'm giving the move to VS 2017 RC a go, could you tell me to which account is the `octokit` organisation linked to in AppVeyor so I can ask them to enable VS 2017 for us? Cheers

**Comment by shiftkey** [comment]
@mderriey it's the `github-windows` account on Appveyor, and if it's something available through the admin panel I can set that for you directly...

**Comment by ghuntley** [comment]
@shiftkey go post in the appveyor thread linked above and once added you can select from admin panel.

**Comment by mderriey** [comment]
@shiftkey Cheers. I don't think it's available in the UI for now as there's a dedicated thread for people to ask that it gets enabled for their accounts. I'll do it. Thanks again.

**Comment by mderriey** [comment]
I have a WIP branch named [port-to-dotnet-core-vs-2017](https://github.com/mderriey/octokit.net/tree/port-to-dotnet-core-vs-2017). I'm waiting for AppVeyor to enable the VS 2017 RC image on the account and I'll create a PR to see if things are running smoothly.

A couple of issues I encountered on the way:

 - `.nuspec` files don't exist anymore and the NuGet package metadata is now stored [in the .csproj](https://docs.nuget.org/ndocs/schema/msbuild-targets#pack-target) so I had to hardcode the current version in there. We could work around that but I went the easy way until we see if we continue on this path
 - MSBuild chokes when the NuGet package version contains `-*` as this was done [here](https://github.com/mderriey/octokit.net/blob/port-to-dotnet-core/build/Tasks/UpdateVersionInfo.cs). I removed it for now
 - Since the new `dotnet` cli doesn't recognise `project.json` files anymore, I transformed `tools.project.json` which we use to install `GitVersion` to [an empty `.csproj`](https://github.com/mderriey/octokit.net/blob/port-to-dotnet-core-vs-2017/build/tools/tools.csproj). I have no idea if that's the way to go

Any feedback is welcome!
Thanks
